# <font color='gold'>**_Análise Exploratória Descritiva: Metacritc Score - Video Games_**</font> 🎮


<kbd><img src="https://raw.githubusercontent.com/JoSEPHDev2022/Metacritc_Scores_Video_Games/main/images/gamer_stock.jpg" width=1000 height=350></kbd>
---

# <font color='gold'>**_Objetivos_**</font>

---

Nesse projeto, iremos realizar uma análise exploratória descritiva de dados em um [Dataset](https://www.kaggle.com/datasets/patkle/metacritic-scores-for-games-movies-tv-and-music) que contém avaliações fornecidas pelo [Metacritic](https://www.metacritic.com/game) sobre diversos títulos de Video-Game. Como um sumário do que será realizado nesse projeto, temos:

<font color='gold'>**_1._**</font> **Compreender os Dados**:
- Aqui será realizada uma análise estatística-descritiva dos dados crus, buscando compreender seu formato e suas características centrais originais.

<font color='gold'>**_2._**</font> **Limpeza, Correção e Tratamento**:
- Realizaremos a limpeza e tratamento desses dados. Com as informações adiquiridas na etapa anterior, iremos buscar adaptar e refatorar o Dataset para o melhor formato possível.

- Utilizaremos funções como principal maneira de limpeza e tratamento.

- Também iremos utilizar de Histogramas e Boxplots/Violinplots para visualizarmos a distribuição dos dados, outliers e etc.

<font color='gold'>**_3._**</font> **Análise Exploratória e Visualizações**:
- Após a compreensão e tratamento desses dados, iremos realizar uma análise exploratória, buscando gerar insights e a retirada de informações relevantes e curiosas dos dados.

- Para isso, utilizaremos de diversas visualizações utilizando Matplotlib e Seaborn.

# <font color='gold'>**_Importando Pacotes e Compreendendo Dados_**</font>

---

In [146]:
# Importando bibliotecas necessárias para análise:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from platform import python_version

# Setando estilo escuto do grid para gráficos Seaborn:
sns.set_theme(context='notebook', style='darkgrid')

# Filtrando avisos de erro:
warnings.filterwarnings('ignore')

# Mostrando versões utilizadas:
print(' Pacotes Prontos! Versões Utilizadas: '.center(70, '-'))
print(f'Python:     --> {python_version()}')
print(f'Pandas:     --> {pd.__version__}')
print(f'Numpy:      --> {np.__version__}')
print(f'Matplotlib: --> {matplotlib.__version__}')
print(f'Seaborn:    --> {sns.__version__}')


---------------- Pacotes Prontos! Versões Utilizadas: ----------------
Python:     --> 3.9.16
Pandas:     --> 1.5.2
Numpy:      --> 1.23.5
Matplotlib: --> 3.6.2
Seaborn:    --> 0.11.2


In [147]:
# Criando o Dataframe:
data_raw = pd.read_csv('data/games.csv', sep=',')

In [148]:
# Visualizando primeiros registros:
data_raw.head(10)

id  metascore       platform       release_date  sort_no  \
0  543718         91             PC    August 18, 2020      301   
1  555108         91             PC   December 8, 2022      302   
2  106820         91  PlayStation 2   November 7, 2005      303   
3  110775         91             PC  November 13, 2008      304   
4  142864         91            Wii   October 26, 2010      305   
5  114951         91            Wii    August 24, 2009      306   
6  397035         91             PC       May 23, 2016      307   
7  102812         91             PC      June 24, 2003      308   
8  161540         91       GameCube   October 23, 2002      309   
9  100705         91  PlayStation 2   October 23, 2000      310   

                                             summary  \
0  From light planes to wide-body jets, fly highl...   
1  Take up your sword, channel your magic or boar...   
2  Strap on your Guitar Hero SG controller, plug-...   
3  Players last visited Northrend in "Warcraft II...   
4  Rock Band returns with the third iteration of ...   
5  Metroid Prime 3: Corruption set a new standard...   
6  Overwatch is a highly stylized team-based shoo...   
7  Cutting-edge technology brings you the most ac...   
8  Drop into the ultimate challenge.  In the next...   
9  EA Sports steps up to the challenge of Sega's ...   

                                       title user_score  
0                 Microsoft Flight Simulator        7.1  
1                             Chained Echoes        8.7  
2                                Guitar Hero        8.5  
3  World of Warcraft: Wrath of the Lich King        7.7  
4                                Rock Band 3        6.8  
5                      Metroid Prime Trilogy          9  
6                                  Overwatch        6.4  
7                       F1 Challenge '99-'02        7.7  
8                   Tony Hawk's Pro Skater 4        8.2  
9                            Madden NFL 2001        7.5

In [149]:
# Visualizando um sample aleatório dos dados:
data_raw.sample(10)

id  metascore       platform        release_date  sort_no  \
10954  145639         71  PlayStation 3       June 18, 2010    10955   
383    104102         92             PC      March 28, 2005      233   
12638  212901         69             PC        June 1, 2012    12639   
5270   558690         79  PlayStation 4         May 5, 2020     5271   
13563  587231         67  PlayStation 5  September 22, 2022    13564   
18189  115453         53            Wii    October 27, 2009    18190   
13727  161919         67           Xbox  September 14, 2003    13728   
7118   182082         77  PlayStation 3    October 16, 2011     7119   
2067   568443         84  Xbox Series X   November 10, 2020     2068   
18372  102259         53  PlayStation 2      March 28, 2002    18373   

                                                 summary  \
10954  SBK 09 is the official game of the Superbike W...   
383    The year is 2008. Information Warfare has evol...   
12638  Home is a unique horror adventure set in a bea...   
5270   Saints & Sinners is a game unlike any other in...   
13563  Inspired by the golden era of skateboarding, S...   
18189  With a playful new art style, SimAnimals Afric...   
13727  Freestyle Metal X unleashes anarchy with the f...   
7118   The small purple dragon returns, bringing with...   
2067   Devil May Cry is back. The legendary stylish a...   
18372  Place your bait, set your hook and reel in the...   

                                          title user_score  
10954       SBK-09 Superbike World Championship        7.4  
383    Tom Clancy's Splinter Cell: Chaos Theory        8.7  
12638                               Home (2012)        5.9  
5270         The Walking Dead: Saints & Sinners        8.2  
13563                        Session: Skate Sim          4  
18189                         SimAnimals Africa        6.3  
13727                          Freestyle MetalX        6.7  
7118              Skylanders: Spyro's Adventure        5.2  
2067           Devil May Cry 5: Special Edition        7.7  
18372                                Top Angler        tbd

Com a visualização geral dos dados podemos nos aprofundar mais em suas características:

In [150]:
# Checando dimensões dos dados:
data_raw.shape

(20022, 8)

Temos um dataset relativamente grande, contendo <font color='gold'>**_20.022_ registros**</font> em <font color='gold'>**_8 variáveis_**</font> diferentes.

Vamos checar informações mais específicas acerca dessas variáveis:

In [151]:
# Visão mais detalhada das variáveis:
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20022 entries, 0 to 20021
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            20022 non-null  int64 
 1   metascore     20022 non-null  int64 
 2   platform      20022 non-null  object
 3   release_date  20022 non-null  object
 4   sort_no       20022 non-null  int64 
 5   summary       19930 non-null  object
 6   title         20022 non-null  object
 7   user_score    20022 non-null  object
dtypes: int64(3), object(5)
memory usage: 1.2+ MB


Informações relevantes:

- Temos alguns valores faltantes apenas na coluna **`summary`**, o que indica que o Dataset está bem completo. Também, esse fator não é um problema por a coluna **`summary`** não será utilizada nas análises.

- Sobre os Data types utilizados: 

    - iremos avaliar mudanças possíveis, como int32 em vez de int64 nas colunas que possuem esse data type. 

    - A coluna **`release_date`** será trasnformada em uma coluna do tipo date em vez de object.

    - A coluna **`user_score`** está no formato object quando deveria estar em float, por se tratar da variável que contém as avaliações fornecidas pelos usuários.
    
Com essas mudanças, a memória consumida será reduzida. Nesse dataset não fará uma diferença notável na performance, porém é uma boa prática sempre deixar os dados no formato mais eficiente possível.

## **Análise Estatística - Descritiva dos Dados**

Conseguimos coletar informações relevantes sobre os dados na etapa anterior, porém para podermos ter uma visão realmente aprofundada desses dados, vamos realizar uma análise estatística para descrever esse conjunto.

- Resumo Estatístico das Variáveis Numéricas:

In [152]:
# Utilizando método describe nas colunas numéricas:
data_raw.describe()

id     metascore       sort_no
count   20022.000000  20022.000000  20022.000000
mean   296058.651084     70.758566  10011.500000
std    190522.456059     12.165042   5779.997881
min    100002.000000     11.000000      1.000000
25%    112180.250000     64.000000   5006.250000
50%    188870.500000     72.000000  10011.500000
75%    510982.250000     80.000000  15016.750000
max    626054.000000     99.000000  20022.000000

Nesse dataset temos três variáveis numéricas, porém apenas **`metascore`** nos importa para essa análise, pois as outras duas variáveis são **`id`** e **`sort_no`** não contém informações relevantes por se tratarem de variáveis de organização dos dados (id, clasificação, etc).

Sobre a coluna **`metascore`**:

- A média geral da avaliação do Metacritic para os jogos (lembrando que essa avaliação vai de 0 a 100) é de <font color='gold'>**70.8**</font>, uma média bem alta considerando o limite de 100.

- O desvio padrão dessas avaliações é de  <font color='gold'>**12.2**</font> pontos acima ou abaixo da média.

- A avaliação mais baixa presente nos dados é de  <font color='gold'>**11**</font>, realmente um valor bem baixo.

- Já a avaliação mais alta é de  <font color='gold'>**99**</font>, aparentemente nenhum game recebeu a nota máxima de 100.

Sobre os quartis e percentis:

- 25% das avaliações do Metacritic estão abaixo ou são iguais a  <font color='gold'>**64**</font>;

- 50% das avaliações do Metacritic estão abaixo ou são iguais a  <font color='gold'>**72**</font>, esse valor representa a mediana das avaliações, ou seja, o valor central dos dados ordenados é 72;

- 75% das avaliações do metacritic estão abaixo ou são iguais a  <font color='gold'>**80**</font>;

- resumo Estatístico das Variáveis Categóricas:

In [153]:
# Utilizando método describe nas colunas categóricas:
data_raw.describe(exclude='number')

platform       release_date  \
count     20022              20022   
unique       22               4794   
top          PC  November 14, 2006   
freq       5276                 48   

                                                  summary          title  \
count                                               19930          20022   
unique                                              15545          12926   
top     Iron Man, an explosive third-person action sho...  Madden NFL 07   
freq                                                    7              9   

       user_score  
count       20022  
unique         94  
top           tbd  
freq         1382

Das variáveis categóricas, **`release_date`** e **`user_score`** não são apropriadas para essa análise, pois deveriam estar em formato date e float respectivamente. Sendo assim, vamos olhar para as variáveis realmente categóricas nos dados:

**`platform`**

- Temos <font color='gold'>**22 plataformas diferentes**</font> nos dados. Isso indica uma variedade grande nas tecnologias utilziadas para games como consoles ao longo do tempo.

- A plataforma em que mais jogos estão presentes é o  <font color='gold'>**PC**</font>, contando com  <font color='gold'>**5.276 registros**</font> nos dados.

**`title`**

- Temos <font color='gold'>**12.926 títulos únicos nos dados**</font>. Isso indica que dos 20.022 títulos totais, 7.096 deles podem ser o mesmo jogo porém em um registro diferente para cada plataforma em que esse jogo foi lançado. Isso será explorado mais a frente.

Como dito previamente, a coluna `summary` não será utilizada, então sua descrição estatística não nos importa.

# <font color='gold'>**_Tratamento e Limpeza dos Dados_**</font>

---

Como analisamos acima com a análise estatística o Dataset não está num formato ideal, tendo data types indevidos para certas variáveis. Esse fator deixa a busca por outliers e a visualização da distribuição dos dados prejudicados, assim, iremos tratar os Data types dos dados, inserir colunas adicionais caso necessário para aprofundar a análise, renomear algumas colunas e reorganizar a distribuição dessas variáveis no Dataset. 

Após esses ajustes, poderemos utilziar de Histogramas e Boxplots para a visualização da distribuição dos dados com seus data types corrigidos, possibilitando a identificação e tratamento de outliers e a compreensão da distribuição desses dados.

Antes de qualquer modificação, se faz necessária a criação de uma cópia dos dados originais, onde será nessa cópia que todas as modificações serão realizadas.

In [154]:
# Criando cópia para modificação dos dados:
data_clean = data_raw.copy()

## **Exluíndo Variávies Desnecessárias para Análise**

As variáveis **`id`**, **`sort_no`** e **`summary`** serão removidas do dataset por não serem necessárias para a composição das análises.

In [155]:
# Excluíndo variáveis:
data_clean = data_clean.drop(['id', 'sort_no', 'summary'], axis=1)

In [156]:
# Verificando dados:
data_clean.head(5)

metascore       platform       release_date  \
0         91             PC    August 18, 2020   
1         91             PC   December 8, 2022   
2         91  PlayStation 2   November 7, 2005   
3         91             PC  November 13, 2008   
4         91            Wii   October 26, 2010   

                                       title user_score  
0                 Microsoft Flight Simulator        7.1  
1                             Chained Echoes        8.7  
2                                Guitar Hero        8.5  
3  World of Warcraft: Wrath of the Lich King        7.7  
4                                Rock Band 3        6.8

## **Renomeando Variáveis**

Com apenas as colunas que queremos presentes nos dados, vamos renomeá-las para padronização e visualização melhores.

In [157]:
# Renomenado variáveis:
data_clean = (data_clean
              .rename(columns=
                      {'metascore': 'Metascore',
                       'platform': 'Platform',
                       'release_date': 'Release_date',
                       'title': 'Game_title',
                       'user_score': 'User_score'}
                       ))

In [158]:
# Visualizando dados:
data_clean.head(5)

Metascore       Platform       Release_date  \
0         91             PC    August 18, 2020   
1         91             PC   December 8, 2022   
2         91  PlayStation 2   November 7, 2005   
3         91             PC  November 13, 2008   
4         91            Wii   October 26, 2010   

                                  Game_title User_score  
0                 Microsoft Flight Simulator        7.1  
1                             Chained Echoes        8.7  
2                                Guitar Hero        8.5  
3  World of Warcraft: Wrath of the Lich King        7.7  
4                                Rock Band 3        6.8

## **Alteração de Data Types**

Agora, vamos alterar os data types necessários. Para isso, vamos checar novamente esses dtypes:

In [159]:
# Checando dtypes:
data_clean.dtypes

Metascore        int64
Platform        object
Release_date    object
Game_title      object
User_score      object
dtype: object

Alterações:

- **`Metascore`** - Transformada para int8. Lembrando que int8 aceita valores que estão entre -128 e 128, já que o Metascore vai de 0 a 100, o formato int8 se torna ideal para essa variável.

- **`Release_date`** - Será transformada em Datetime.

- **`User_score`** - Transformada para float.

Para termos certeza que nossas alterações serão realizadas, vamos checar se a coluna **`Metascore`** e **`User_score`** possuem apenas valores numéricos:

In [160]:
# Checando se colunas possuem apenas valores numéricos:
print('Variável         Possuí Apenas Numéricos')

data_clean.apply(lambda values: 
                    pd.to_numeric(values, errors='coerce').notnull().all())

Variável         Possuí Apenas Numéricos


Metascore        True
Platform        False
Release_date    False
Game_title      False
User_score      False
dtype: bool

Com isso podemos notar que a variável **`Metascore`** possui apenas valores numéricos, porém **`User_score`** não. Sendo assim, para podermos alterar o dtype dessa variável, precisamos antes remover valores que não são numéricos. Vamos checar essa variável e seus valores:

In [161]:
# Checando valores da coluna user_score:
data_clean['User_score'].unique()

array(['7.1', '8.7', '8.5', '7.7', '6.8', '9', '6.4', '8.2', '7.5', '8.9',
       '9.1', '8.1', '8.6', '8.3', '7.9', '9.6', '8', '8.4', '5.9', '7.4',
       '6.7', '6.6', '8.8', '6', '7.8', '9.2', '4', '9.5', '6.3', '7.6',
       '7.3', '7.2', '6.2', '9.3', '7', '6.5', '5.6', '6.9', '5.7', '5.4',
       '3.3', '5', '5.2', '9.4', '5.8', '5.3', '3.5', '4.8', '6.1', '4.7',
       '4.2', '3.4', '2.1', 'tbd', '4.5', '4.1', '5.5', '3', '2.8', '4.4',
       '3.9', '4.9', '3.6', '4.3', '5.1', '1.8', '4.6', '3.1', '9.7',
       '3.2', '2.2', '2.9', '2.5', '3.7', '3.8', '2', '2.7', '2.3', '1.6',
       '1.3', '2.6', '0.9', '2.4', '1.4', '1.5', '1.7', '1', '1.1', '1.9',
       '0.6', '0.5', '0.3', '1.2', '0.7'], dtype=object)

Com isso, vemos que alguns registros possuem o valor **'tbd'** que é uma string. Vamos checar quantos registros possuem essa string:

In [162]:
# Contando aparições da string 'tbd':
data_clean[data_clean['User_score'] == 'tbd']['User_score'].count() 

1382

Temos **1.382** registros que contém a string 'tbd'. Com uma rápida pesquisa na página de fonte desses dados no Metacritic, temos que a string 'tbd' representa jogos que não tiveram nenhuma avaliação dos usuários até então. Nesse caso, iremos substituir esses valores por 0 antes de realizar a alteração do data type para float.

In [163]:
# Substituíndo a string por 0:
data_clean['User_score'] = (data_clean['User_score']
              .apply(lambda value: 0 if value == 'tbd' else value))

In [164]:
# Contando aparições da string 'tbd':
data_clean[data_clean['User_score'] == 'tbd']['User_score'].count() 

0

Com as variáveis no formato adequado, podemos alterar os data types:

In [165]:
# Alterando dtypes:
data_clean = (data_clean
              .astype({'Metascore': 'int8',
                       'User_score': 'float'}))

data_clean['Release_date'] = pd.to_datetime(data_clean['Release_date'])

Com as alterações feitas nos data types, vamos checar o consumo de memória dos dados:

In [166]:
data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20022 entries, 0 to 20021
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Metascore     20022 non-null  int8          
 1   Platform      20022 non-null  object        
 2   Release_date  20022 non-null  datetime64[ns]
 3   Game_title    20022 non-null  object        
 4   User_score    20022 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int8(1), object(2)
memory usage: 645.4+ KB


Podemos ver a diferença expressiva que as alterações de data types teve na performance dos nossos dados. Antes da alteração dos dados, tinhamos um consumo de memória de 1.2+ MB, agora com as alterações, temos um uso de memória de 528+ KB.

In [167]:
# Checando dados:
data_clean.head(5)

Metascore       Platform Release_date  \
0         91             PC   2020-08-18   
1         91             PC   2022-12-08   
2         91  PlayStation 2   2005-11-07   
3         91             PC   2008-11-13   
4         91            Wii   2010-10-26   

                                  Game_title  User_score  
0                 Microsoft Flight Simulator         7.1  
1                             Chained Echoes         8.7  
2                                Guitar Hero         8.5  
3  World of Warcraft: Wrath of the Lich King         7.7  
4                                Rock Band 3         6.8

## **Criando Nova Variável: Trimestre de Lançamento**

Como forma de expandir nossas possibilidades de análise, vamos inserir nos dados uma nova variável que informa qual o trimestre do ano em que o Jogo foi lançado.

In [168]:
def set_year_quarter(data: pd.DataFrame) -> pd.DataFrame:
    '''Retorna uma nova variável que informa o trimestre do ano em que
    o jogo foi lançado. Essa definição de trimestre se baseia no mês de lançamento
    da coluna Release_date.
    
    PARÂMETROS:
    
        data > DataFrame para se aplicar a alteração.
    '''
    data['Release_quarter'] = data['Release_date'].apply(
        lambda value:
            'Q1' if value.month in [1, 2, 3] else
            'Q2' if value.month in [4, 5, 6] else
            'Q3' if value.month in [7, 8, 9] else
            'Q4' if value.month in [10, 11, 12] else
            value
            )
    
    return data

In [169]:
# Aplicando função:
data_clean = set_year_quarter(data_clean)

# Checando resultado:
data_clean.sample(10)

Metascore       Platform Release_date  \
8613          74             PC   2019-01-15   
6470          77  PlayStation 3   2010-10-19   
15369         63         Switch   2018-04-20   
6070          78             PC   2021-07-28   
16051         61             PC   2011-05-30   
11307         71         Switch   2018-04-26   
14733         65             PC   2017-11-17   
3516          82             PC   2009-04-13   
9136          74         Switch   2019-10-08   
4294          80             PC   2022-11-16   

                                              Game_title  User_score  \
8613   The Walking Dead: The Telltale Series - The Fi...         7.7   
6470                                       EA Sports MMA         6.7   
15369                                 The Way Remastered         6.9   
6070                               Unbound: Worlds Apart         7.6   
16051                          Hunted: The Demon's Forge         6.6   
11307                               A Robot Named Fight!         7.5   
14733                           Star Wars Battlefront II         1.9   
3516                                    DCS: Black Shark         8.5   
9136                    The Alliance Alive HD Remastered         7.3   
4294                           Call of Duty: Warzone 2.0         2.3   

      Release_quarter  
8613               Q1  
6470               Q4  
15369              Q2  
6070               Q3  
16051              Q2  
11307              Q2  
14733              Q4  
3516               Q2  
9136               Q4  
4294               Q4

## **Criando Nova Variável: Subdivisões da Variável Release_date**

Com a nova variável de trimestre de lançamento criada, vamos retornar a variável **`Release_date`** e utilizá-la para criar três variáveis distintas que informam apenas o dia, mês e ano do lançamento.

Essa divisão da variável **`Release_date`** será feita para facilitar a criação de plots na etapa de análises.

- Criando variável **`Release_day`**:

In [170]:
# Criando nova variável:
data_clean['Release_day'] = data_clean['Release_date'].dt.day

- Criando variável **`Release_month`**:

In [171]:
# Criando nova variável:
data_clean['Release_month'] = data_clean['Release_date'].dt.month

- Criando variável **`Release_year`**:

In [172]:
# Criando nova variável:
data_clean['Release_year'] = data_clean['Release_date'].dt.year

In [173]:
# Checando dados:
data_clean.head(5)

Metascore       Platform Release_date  \
0         91             PC   2020-08-18   
1         91             PC   2022-12-08   
2         91  PlayStation 2   2005-11-07   
3         91             PC   2008-11-13   
4         91            Wii   2010-10-26   

                                  Game_title  User_score Release_quarter  \
0                 Microsoft Flight Simulator         7.1              Q3   
1                             Chained Echoes         8.7              Q4   
2                                Guitar Hero         8.5              Q4   
3  World of Warcraft: Wrath of the Lich King         7.7              Q4   
4                                Rock Band 3         6.8              Q4   

   Release_day  Release_month  Release_year  
0           18              8          2020  
1            8             12          2022  
2            7             11          2005  
3           13             11          2008  
4           26             10          2010

## **Criando Nova Variável: Metascore_class**

Essa variável tem como objetivo abrigar a classificação dos valores do Metascore. De acordo com um [FAQ do próprio site](https://www.metacritic.com/about-metascores), para cada range de notas temos uma classificação geral atribuída a ele como ilustrado na imagem abaixo retirada diretamente do FAQ:

<img src="https://raw.githubusercontent.com/JoSEPHDev2022/Metacritc_Scores_Video_Games/main/images/metascore_classification.png" width=700 height=350>

Com essa nova variável, teremos mais possibilidades de análise posteriores.

In [174]:
def set_metascore_class(data: pd.DataFrame) -> pd.DataFrame:
    '''Retorna uma nova variável que consta a classificação do Jogo
    com base em sua nota. Essas classificações são:
    
        - "Universal Acclaim" para games com notas entre 90 e 100;
        - "Generally Favorable Reviews" para games com notas entre 75 e 89;
        - "Mixed or Average Reviews" para games com notas entre 50 e 74;
        - "Generally Unfavorable Reviews" para games com notas entre 20 e 49;
        - "Overwhelming Dislike" para games com notas entre 0 e 19.

        PARÂMETROS:
        
        data > DataFrame para se aplicar a alteração. 
    '''
    data['Classification'] = data['Metascore'].apply(
        lambda value:
            'Universal Acclaim' if 90 <= value <= 100 else
            'Generally Favorable Reviews' if 75 <= value <= 89 else
            'Mixed or Average Reviews' if 50 <= value <= 74 else
            'Generally Unfavorable Reviews' if 20 <= value <= 49 else
            'Overwhelming Dislike'   
            )
    
    return data

In [175]:
# Aplicando função:
data_clean = set_metascore_class(data_clean)

In [176]:
# Visualizando resultados:
data_clean.sample(10)

Metascore          Platform Release_date  \
2956          83          Xbox 360   2008-07-15   
6382          78                PC   2016-04-28   
2723          83  PlayStation Vita   2016-06-28   
6693          77                PC   2013-12-13   
12643         69              Xbox   2004-06-16   
17682         56     PlayStation 4   2020-09-22   
2554          83     PlayStation 4   2019-08-20   
11369         71     PlayStation 4   2016-10-25   
4846          80     PlayStation 4   2014-03-21   
1026          87     PlayStation 4   2019-10-11   

                                  Game_title  User_score Release_quarter  \
2956                        NCAA Football 09         7.7              Q3   
6382                Offworld Trading Company         7.6              Q2   
2723          Zero Escape: Zero Time Dilemma         8.3              Q2   
6693                               Teslagrad         7.8              Q4   
12643                             Powerdrome         7.0              Q2   
17682                    Tennis World Tour 2         6.3              Q3   
2554        The Yakuza Remastered Collection         7.8              Q3   
11369  Sherlock Holmes: The Devil's Daughter         6.8              Q4   
4846                    inFamous: Second Son         8.0              Q1   
1026              Frostpunk: Console Edition         8.0              Q4   

       Release_day  Release_month  Release_year               Classification  
2956            15              7          2008  Generally Favorable Reviews  
6382            28              4          2016  Generally Favorable Reviews  
2723            28              6          2016  Generally Favorable Reviews  
6693            13             12          2013  Generally Favorable Reviews  
12643           16              6          2004     Mixed or Average Reviews  
17682           22              9          2020     Mixed or Average Reviews  
2554            20              8          2019  Generally Favorable Reviews  
11369           25             10          2016     Mixed or Average Reviews  
4846            21              3          2014  Generally Favorable Reviews  
1026            11             10          2019  Generally Favorable Reviews

## **Reorganizando Sequência de Variáveis**

Com todas as variáveis com data types apropriados e com a inserção das novas variáveis, basta organizar os dados para uma visualização melhor da tabela:

In [177]:
# Reorganizando aparição das colunas no DataFrame:
data_clean = data_clean.loc[:, 
                            ['Game_title', 'Platform', 'Release_date', 'Release_year', 
                             'Release_month', 'Release_day', 'Release_quarter', 'User_score', 
                             'Metascore', 'Classification']]

In [179]:
# Verificando forma final dos dados:
data_clean.head(10)

Game_title       Platform Release_date  \
0                 Microsoft Flight Simulator             PC   2020-08-18   
1                             Chained Echoes             PC   2022-12-08   
2                                Guitar Hero  PlayStation 2   2005-11-07   
3  World of Warcraft: Wrath of the Lich King             PC   2008-11-13   
4                                Rock Band 3            Wii   2010-10-26   
5                      Metroid Prime Trilogy            Wii   2009-08-24   
6                                  Overwatch             PC   2016-05-23   
7                       F1 Challenge '99-'02             PC   2003-06-24   
8                   Tony Hawk's Pro Skater 4       GameCube   2002-10-23   
9                            Madden NFL 2001  PlayStation 2   2000-10-23   

   Release_year  Release_month  Release_day Release_quarter  User_score  \
0          2020              8           18              Q3         7.1   
1          2022             12            8              Q4         8.7   
2          2005             11            7              Q4         8.5   
3          2008             11           13              Q4         7.7   
4          2010             10           26              Q4         6.8   
5          2009              8           24              Q3         9.0   
6          2016              5           23              Q2         6.4   
7          2003              6           24              Q2         7.7   
8          2002             10           23              Q4         8.2   
9          2000             10           23              Q4         7.5   

   Metascore     Classification  
0         91  Universal Acclaim  
1         91  Universal Acclaim  
2         91  Universal Acclaim  
3         91  Universal Acclaim  
4         91  Universal Acclaim  
5         91  Universal Acclaim  
6         91  Universal Acclaim  
7         91  Universal Acclaim  
8         91  Universal Acclaim  
9         91  Universal Acclaim